mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp

```diff
@@ -0,0 +1,112 @@
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/SparseTensor/IR/SparseTensor.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+using namespace std;
+using namespace mlir;
+using namespace mlir::sparse_tensor;
+
+// make CSR tensor type
+mlir::RankedTensorType getCSRTensorType(MLIRContext *context, Type valueType)
+{
+    SmallVector<sparse_tensor::SparseTensorEncodingAttr::DimLevelType, 2> dlt;
+    dlt.push_back(sparse_tensor::SparseTensorEncodingAttr::DimLevelType::Dense);
+    dlt.push_back(sparse_tensor::SparseTensorEncodingAttr::DimLevelType::Compressed);
+    unsigned ptr = 64;
+    unsigned ind = 64;
+    AffineMap map = AffineMap::getMultiDimIdentityMap(2, context);
+
+    RankedTensorType csrTensor = RankedTensorType::get(
+        {-1, -1}, /* 2D, unknown size */
+        valueType,
+        sparse_tensor::SparseTensorEncodingAttr::get(context, dlt, map, ptr, ind));
+
+    return csrTensor;
+}
+
+/// Returns function reference (first hit also inserts into module).
+// from: llvm/llvm-project/mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp
+static FlatSymbolRefAttr getFunc(mlir::ModuleOp &mod, mlir::Location &loc, StringRef name, TypeRange result,
+                                 TypeRange operands)
+{
+    MLIRContext *context = mod.getContext();
+    auto func = mod.lookupSymbol<FuncOp>(name);
+    if (!func)
+    {
+        OpBuilder moduleBuilder(mod.getBodyRegion());
+        moduleBuilder
+            .create<FuncOp>(loc, name,
+                            FunctionType::get(context, operands, result))
+            .setPrivate();
+    }
+    return SymbolRefAttr::get(context, name);
+}
+
+mlir::CallOp callEmptyLike(mlir::OpBuilder &builder, mlir::ModuleOp &mod, mlir::Location loc, mlir::Value tensor) {
+    auto tensorType = tensor.getType();
```

Comment:
While it's nice to use `auto`, I like making the types explicit since it'll be easier to look up in the MLIR doxygen docs what methods are available for a certain value. 

For example, if I knew that `tensorType` was a `mlir::Type`, I could find at https://mlir.llvm.org/doxygen/classmlir_1_1Type.html what methods were available for it. 

There's also some implicit knowledge here in labelling it as `mlir::Type` (instead of `mlir::TensorType`, which is what it would actually always be in this function) telling us that the `getType` method returns a general `mlir::Type` and that we'd have to dynamically cast it to a `mlir::TensorType` if we wanted to use methods specific to `mlir::TensorType`. 
- In short, it's nice to know that the compiler sees this as a `mlir::Type` and not a `mlir::TensorType` so that we'd know we have to dynamically cast it if we wanted to use the methods of `mlir::TensorType`.

A silly hack that I use to determine the type of a variable is to print it (which'll cause a compile-time type error), compile the code, and read the error message of "Expected X, but got Y" to know that the variable's type should be Y. This won't work if it is printable, but that's rarely the case.

Reply:
Good point.  I made this change to the pass but not here.  Will fix.